src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonArrayConverter.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections;
+using System.Collections.Generic;
+
+namespace System.Text.Json.Serialization.Converters
+{
+    /// <summary>
+    /// Converter for <cref>System.Array</cref>.
+    /// </summary>
+    internal sealed class JsonArrayConverter<TCollection, TElement>
+        : JsonIEnumerableDefaultConverter<TCollection, TElement>
+        where TCollection: IEnumerable
+    {
+        internal override bool CanHaveIdMetadata => false;
+
+        protected override void Add(TElement value, ref ReadStack state)
+        {
+            ((List<TElement>)state.Current.ReturnValue!).Add(value);
+        }
+
+        protected override void CreateCollection(ref ReadStack state, JsonSerializerOptions options)
+        {
+            state.Current.ReturnValue = new List<TElement>();
+        }
+
+        protected override void ConvertCollection(ref ReadStack state, JsonSerializerOptions options)
+        {
+            List<TElement> list = (List<TElement>)state.Current.ReturnValue!;
+            state.Current.ReturnValue = list.ToArray();
+        }
+
+        protected override bool OnWriteResume(Utf8JsonWriter writer, TCollection value, JsonSerializerOptions options, ref WriteStack state)
+        {
+            TElement[] array = (TElement[])(IEnumerable)value;
```

Comment:
Yes, this converter is only used for arrays. The selection logic in the IEnumerable factory needs to stay correct.

Reply:
Yes this converter is only selected when the property is an array. Unfortunately there is no support for generic constraints with arrays, otherwise I would have done that.

I will add an Assert for the cast.